// devframe/dist/recipes/interactive-auth.mjs
import { n as colors } from "../diagnostics-reporter-CsIG85Q5.mjs";
import { isAnonymousRpcMethod } from "../constants.mjs";
import { n as defineRpcFunction } from "../define-BLWPsH6y.mjs";
import { t as getInternalContext } from "../context-CKJheEtf.mjs";
import { a as verifyAuthToken, n as exchangeTempAuthCode, r as getTempAuthCode, t as buildOtpAuthUrl } from "../state-WX3HT5a3.mjs";
import * as v from "valibot";
//#region src/recipes/interactive-auth.ts
function defaultBanner(info) {
	console.log(`\n  ${colors.dim("devframe auth code")}  ${colors.bold(info.code)}\n  ${colors.dim("or open")}       ${colors.cyan(info.url)}\n`);
}
/**
* Package the interactive OTP auth protocol devframe's primitives
* (`exchangeTempAuthCode`, `verifyAuthToken`, `revokeAuthToken`,
* `getTempAuthCode`, `buildOtpAuthUrl`) implement into a ready-made
* {@link DevframeAuthHandler} — the handshake RPC functions, the resolver
* gate, the connect-time trust hook, and the startup banner.
*
* The auth storage stays internal to this handler — callers never reach into
* `devframe/node/hub-internals` themselves.
*
* ```ts
* import { createInteractiveAuth } from 'devframe/recipes/interactive-auth'
*
* const auth = createInteractiveAuth(ctx)
* auth.rpcFunctions.forEach(fn => ctx.rpc.register(fn))
* auth.printBanner()
*
* // wire `auth.authorize` / `auth.onConnect` into your transport, or pass
* // the whole handler to `startHttpAndWs({ auth, ... })`.
* ```
*/
function createInteractiveAuth(context, options = {}) {
	const internal = getInternalContext(context);
	const storage = internal.storage.auth;
	const staticTokens = new Set(options.clientAuthTokens ?? []);
	function isStaticToken(token) {
		return !!token && staticTokens.has(token);
	}
	function resolveServerUrl() {
		return options.serverUrl?.() ?? context.host.resolveOrigin();
	}
	let bannerPrintedForCode;
	function printBanner() {
		const code = getTempAuthCode();
		if (code === bannerPrintedForCode) return;
		bannerPrintedForCode = code;
		const url = buildOtpAuthUrl(resolveServerUrl(), code);
		(options.banner ?? defaultBanner)({
			code,
			url
		});
	}
	const anonymousAuth = defineRpcFunction({
		name: "anonymous:devframe:auth",
		type: "action",
		jsonSerializable: true,
		args: [v.object({
			authToken: v.string(),
			ua: v.string(),
			origin: v.string()
		})],
		returns: v.object({ isTrusted: v.boolean() }),
		handler(params) {
			const session = context.rpc.getCurrentRpcSession();
			if (!session) return { isTrusted: false };
			if (isStaticToken(params.authToken)) {
				session.meta.clientAuthToken = params.authToken;
				session.meta.isTrusted = true;
				return { isTrusted: true };
			}
			return { isTrusted: verifyAuthToken(params.authToken, session, storage) };
		}
	});
	const anonymousAuthExchange = defineRpcFunction({
		name: "anonymous:devframe:auth:exchange",
		type: "action",
		jsonSerializable: true,
		args: [v.object({
			code: v.string(),
			ua: v.string(),
			origin: v.string()
		})],
		returns: v.object({ authToken: v.nullable(v.string()) }),
		handler(params) {
			const session = context.rpc.getCurrentRpcSession();
			if (!session) return { authToken: null };
			const authToken = exchangeTempAuthCode(params.code, session, params, storage);
			printBanner();
			return { authToken };
		}
	});
	const revoke = defineRpcFunction({
		name: "devframe:auth:revoke",
		type: "action",
		jsonSerializable: true,
		args: [],
		returns: v.void(),
		async handler() {
			const token = context.rpc.getCurrentRpcSession()?.meta.clientAuthToken;
			if (token) await internal.revokeAuthToken(token);
		}
	});
	function authorize(methodName, session) {
		if (isAnonymousRpcMethod(methodName)) return true;
		return !!session.meta.isTrusted;
	}
	function onConnect(peer, session) {
		let token;
		try {
			token = new URL(peer.request?.url ?? "", "http://localhost").searchParams.get("devframe_auth_token") ?? void 0;
		} catch {}
		if (!token) return;
		if (isStaticToken(token)) {
			session.meta.clientAuthToken = token;
			session.meta.isTrusted = true;
			return;
		}
		verifyAuthToken(token, session, storage);
	}
	return {
		rpcFunctions: [
			anonymousAuth,
			anonymousAuthExchange,
			revoke
		],
		authorize,
		onConnect,
		printBanner
	};
}
//#endregion
export { createInteractiveAuth };
