// @vitejs/devtools/dist/plugins-BeKV7vlV.js
import { t as DevToolsServer } from "./server-tX7Wj9Z1.js";
import { dirDist } from "./dirs.js";
import process$1 from "node:process";
import { colors } from "devframe/utils/colors";
import { join, normalize, resolve } from "pathe";
//#region src/node/plugins/build.ts
function DevToolsBuild(options = {}) {
	let context;
	let resolvedConfig;
	return {
		name: "vite:devtools:build",
		apply: "build",
		configResolved(config) {
			resolvedConfig = config;
		},
		async buildStart() {
			const { createDevToolsContext } = await import("./server-tX7Wj9Z1.js").then((n) => n.r);
			context = await createDevToolsContext(resolvedConfig);
		},
		async closeBundle() {
			console.log(colors.cyan`${"⬢"} Building static Vite DevTools...`);
			const outDir = options.outDir ? resolve(resolvedConfig.root, options.outDir) : resolve(resolvedConfig.root, resolvedConfig.build.outDir);
			const { buildStaticDevTools } = await import("./build-static-v0Ugy0lq.js");
			await buildStaticDevTools({
				context,
				outDir,
				withApp: true
			});
		}
	};
}
//#endregion
//#region src/node/plugins/injection.ts
const DEVTOOLS_INJECTION_VIRTUAL_ID = "virtual:vite-devtools-injection";
const RESOLVED_DEVTOOLS_INJECTION_VIRTUAL_ID = `\0${DEVTOOLS_INJECTION_VIRTUAL_ID}`;
function resolveDevToolsInjectionEntry() {
	return process$1.env.VITE_DEVTOOLS_LOCAL_DEV ? normalize(join(dirDist, "..", "src/client/inject/index.ts")) : normalize(join(dirDist, "client/inject.js"));
}
function DevToolsInjection() {
	return {
		name: "vite:devtools:injection",
		enforce: "post",
		apply(_config, env) {
			return env.command === "serve" && !env.isSsrBuild;
		},
		transformIndexHtml: {
			order: "pre",
			handler() {
				return [{
					tag: "script",
					attrs: { type: "module" },
					children: `import ${JSON.stringify(DEVTOOLS_INJECTION_VIRTUAL_ID)}`,
					injectTo: "body"
				}];
			}
		},
		resolveId(id) {
			if (id === DEVTOOLS_INJECTION_VIRTUAL_ID) return RESOLVED_DEVTOOLS_INJECTION_VIRTUAL_ID;
		},
		load(id) {
			if (id === RESOLVED_DEVTOOLS_INJECTION_VIRTUAL_ID) return `import(${JSON.stringify(resolveDevToolsInjectionEntry())})\n`;
		}
	};
}
//#endregion
//#region src/node/plugins/index.ts
async function DevTools(options = {}) {
	const { builtinDevTools = true, build } = options;
	const plugins = [DevToolsInjection(), DevToolsServer()];
	if (build?.withApp) plugins.push(DevToolsBuild({ outDir: build.outDir }));
	if (builtinDevTools) plugins.push(await import("@vitejs/devtools-rolldown").then((m) => m.DevToolsRolldownUI()));
	return plugins;
}
//#endregion
export { DevTools as t };
