// @vitejs/devtools/dist/server-tX7Wj9Z1.js
import { i as __exportAll, r as __commonJSMin } from "./dist-HTAOCnE-.js";
import { n as diagnostics, t as MARK_INFO } from "./constants-CfdflDK5.js";
import { t as createDevToolsMiddleware } from "./server-CRE2OchS.js";
import { DEVTOOLS_DOCK_IMPORTS_VIRTUAL_ID, DEVTOOLS_MOUNT_PATH, DEVTOOLS_MOUNT_PATH_NO_TRAILING_SLASH, DEVTOOLS_VITEPLUS_GROUP_ID } from "@vitejs/devtools-kit/constants";
import { createKitContext, createViteDevToolsHost } from "@vitejs/devtools-kit/node";
import { isObject } from "devframe/node";
import { createDebug } from "obug";
import process$1, { stdin, stdout } from "node:process";
import { styleText } from "node:util";
import l__default from "node:readline";
import "node:tty";
import "node:fs";
import { relative, resolve } from "node:path";
import { defineRpcFunction } from "@vitejs/devtools-kit";
import { abortPendingAuth, getTempAuthToken, refreshTempAuthToken, setPendingAuth } from "devframe/node/auth";
import { getInternalContext } from "devframe/node/hub-internals";
import { colors } from "devframe/utils/colors";
import { launchEditor } from "devframe/utils/launch-editor";
import { open } from "devframe/utils/open";
//#region ../../node_modules/.pnpm/fast-string-truncated-width@3.0.3/node_modules/fast-string-truncated-width/dist/utils.js
const getCodePointsLength = (() => {
	const SURROGATE_PAIR_RE = /[\uD800-\uDBFF][\uDC00-\uDFFF]/g;
	return (input) => {
		let surrogatePairsNr = 0;
		SURROGATE_PAIR_RE.lastIndex = 0;
		while (SURROGATE_PAIR_RE.test(input)) surrogatePairsNr += 1;
		return input.length - surrogatePairsNr;
	};
})();
const isFullWidth = (x) => {
	return x === 12288 || x >= 65281 && x <= 65376 || x >= 65504 && x <= 65510;
};
const isWideNotCJKTNotEmoji = (x) => {
	return x === 8987 || x === 9001 || x >= 12272 && x <= 12287 || x >= 12289 && x <= 12350 || x >= 12441 && x <= 12543 || x >= 12549 && x <= 12591 || x >= 12593 && x <= 12686 || x >= 12688 && x <= 12771 || x >= 12783 && x <= 12830 || x >= 12832 && x <= 12871 || x >= 12880 && x <= 19903 || x >= 65040 && x <= 65049 || x >= 65072 && x <= 65106 || x >= 65108 && x <= 65126 || x >= 65128 && x <= 65131 || x >= 127488 && x <= 127490 || x >= 127504 && x <= 127547 || x >= 127552 && x <= 127560 || x >= 131072 && x <= 196605 || x >= 196608 && x <= 262141;
};
//#endregion
//#region ../../node_modules/.pnpm/fast-string-truncated-width@3.0.3/node_modules/fast-string-truncated-width/dist/index.js
const ANSI_RE = /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]|\u001b\]8;[^;]*;.*?(?:\u0007|\u001b\u005c)/y;
const CONTROL_RE = /[\x00-\x08\x0A-\x1F\x7F-\x9F]{1,1000}/y;
const CJKT_WIDE_RE = /(?:(?![\uFF61-\uFF9F\uFF00-\uFFEF])[\p{Script=Han}\p{Script=Hiragana}\p{Script=Katakana}\p{Script=Hangul}\p{Script=Tangut}]){1,1000}/uy;
const TAB_RE = /\t{1,1000}/y;
const EMOJI_RE = /[\u{1F1E6}-\u{1F1FF}]{2}|\u{1F3F4}[\u{E0061}-\u{E007A}]{2}[\u{E0030}-\u{E0039}\u{E0061}-\u{E007A}]{1,3}\u{E007F}|(?:\p{Emoji}\uFE0F\u20E3?|\p{Emoji_Modifier_Base}\p{Emoji_Modifier}?|\p{Emoji_Presentation})(?:\u200D(?:\p{Emoji_Modifier_Base}\p{Emoji_Modifier}?|\p{Emoji_Presentation}|\p{Emoji}\uFE0F\u20E3?))*/uy;
const LATIN_RE = /(?:[\x20-\x7E\xA0-\xFF](?!\uFE0F)){1,1000}/y;
const MODIFIER_RE = /\p{M}+/gu;
const NO_TRUNCATION$1 = {
	limit: Infinity,
	ellipsis: ""
};
const getStringTruncatedWidth = (input, truncationOptions = {}, widthOptions = {}) => {
	const LIMIT = truncationOptions.limit ?? Infinity;
	const ELLIPSIS = truncationOptions.ellipsis ?? "";
	const ELLIPSIS_WIDTH = truncationOptions?.ellipsisWidth ?? (ELLIPSIS ? getStringTruncatedWidth(ELLIPSIS, NO_TRUNCATION$1, widthOptions).width : 0);
	const ANSI_WIDTH = 0;
	const CONTROL_WIDTH = widthOptions.controlWidth ?? 0;
	const TAB_WIDTH = widthOptions.tabWidth ?? 8;
	const EMOJI_WIDTH = widthOptions.emojiWidth ?? 2;
	const FULL_WIDTH_WIDTH = 2;
	const REGULAR_WIDTH = widthOptions.regularWidth ?? 1;
	const WIDE_WIDTH = widthOptions.wideWidth ?? FULL_WIDTH_WIDTH;
	const PARSE_BLOCKS = [
		[LATIN_RE, REGULAR_WIDTH],
		[ANSI_RE, ANSI_WIDTH],
		[CONTROL_RE, CONTROL_WIDTH],
		[TAB_RE, TAB_WIDTH],
		[EMOJI_RE, EMOJI_WIDTH],
		[CJKT_WIDE_RE, WIDE_WIDTH]
	];
	let indexPrev = 0;
	let index = 0;
	let length = input.length;
	let lengthExtra = 0;
	let truncationEnabled = false;
	let truncationIndex = length;
	let truncationLimit = Math.max(0, LIMIT - ELLIPSIS_WIDTH);
	let unmatchedStart = 0;
	let unmatchedEnd = 0;
	let width = 0;
	let widthExtra = 0;
	outer: while (true) {
		if (unmatchedEnd > unmatchedStart || index >= length && index > indexPrev) {
			const unmatched = input.slice(unmatchedStart, unmatchedEnd) || input.slice(indexPrev, index);
			lengthExtra = 0;
			for (const char of unmatched.replaceAll(MODIFIER_RE, "")) {
				const codePoint = char.codePointAt(0) || 0;
				if (isFullWidth(codePoint)) widthExtra = FULL_WIDTH_WIDTH;
				else if (isWideNotCJKTNotEmoji(codePoint)) widthExtra = WIDE_WIDTH;
				else widthExtra = REGULAR_WIDTH;
				if (width + widthExtra > truncationLimit) truncationIndex = Math.min(truncationIndex, Math.max(unmatchedStart, indexPrev) + lengthExtra);
				if (width + widthExtra > LIMIT) {
					truncationEnabled = true;
					break outer;
				}
				lengthExtra += char.length;
				width += widthExtra;
			}
			unmatchedStart = unmatchedEnd = 0;
		}
		if (index >= length) break outer;
		for (let i = 0, l = PARSE_BLOCKS.length; i < l; i++) {
			const [BLOCK_RE, BLOCK_WIDTH] = PARSE_BLOCKS[i];
			BLOCK_RE.lastIndex = index;
			if (BLOCK_RE.test(input)) {
				lengthExtra = BLOCK_RE === CJKT_WIDE_RE ? getCodePointsLength(input.slice(index, BLOCK_RE.lastIndex)) : BLOCK_RE === EMOJI_RE ? 1 : BLOCK_RE.lastIndex - index;
				widthExtra = lengthExtra * BLOCK_WIDTH;
				if (width + widthExtra > truncationLimit) truncationIndex = Math.min(truncationIndex, index + Math.floor((truncationLimit - width) / BLOCK_WIDTH));
				if (width + widthExtra > LIMIT) {
					truncationEnabled = true;
					break outer;
				}
				width += widthExtra;
				unmatchedStart = indexPrev;
				unmatchedEnd = index;
				index = indexPrev = BLOCK_RE.lastIndex;
				continue outer;
			}
		}
		index += 1;
	}
	return {
		width: truncationEnabled ? truncationLimit : width,
		index: truncationEnabled ? truncationIndex : length,
		truncated: truncationEnabled,
		ellipsed: truncationEnabled && LIMIT >= ELLIPSIS_WIDTH
	};
};
//#endregion
//#region ../../node_modules/.pnpm/fast-string-width@3.0.2/node_modules/fast-string-width/dist/index.js
const NO_TRUNCATION = {
	limit: Infinity,
	ellipsis: "",
	ellipsisWidth: 0
};
const fastStringWidth = (input, options = {}) => {
	return getStringTruncatedWidth(input, NO_TRUNCATION, options).width;
};
//#endregion
//#region ../../node_modules/.pnpm/fast-wrap-ansi@0.2.0/node_modules/fast-wrap-ansi/lib/main.js
const ESC = "\x1B";
const CSI = "";
const END_CODE = 39;
const ANSI_ESCAPE_BELL = "\x07";
const ANSI_CSI = "[";
const ANSI_OSC = "]";
const ANSI_SGR_TERMINATOR = "m";
const ANSI_ESCAPE_LINK = `${ANSI_OSC}8;;`;
const GROUP_REGEX = new RegExp(`(?:\\${ANSI_CSI}(?<code>\\d+)m|\\${ANSI_ESCAPE_LINK}(?<uri>.*)${ANSI_ESCAPE_BELL})`, "y");
const getClosingCode = (openingCode) => {
	if (openingCode >= 30 && openingCode <= 37) return 39;
	if (openingCode >= 90 && openingCode <= 97) return 39;
	if (openingCode >= 40 && openingCode <= 47) return 49;
	if (openingCode >= 100 && openingCode <= 107) return 49;
	if (openingCode === 1 || openingCode === 2) return 22;
	if (openingCode === 3) return 23;
	if (openingCode === 4) return 24;
	if (openingCode === 7) return 27;
	if (openingCode === 8) return 28;
	if (openingCode === 9) return 29;
	if (openingCode === 0) return 0;
};
const wrapAnsiCode = (code) => `${ESC}${ANSI_CSI}${code}${ANSI_SGR_TERMINATOR}`;
const wrapAnsiHyperlink = (url) => `${ESC}${ANSI_ESCAPE_LINK}${url}${ANSI_ESCAPE_BELL}`;
const wrapWord = (rows, word, columns) => {
	const characters = word[Symbol.iterator]();
	let isInsideEscape = false;
	let isInsideLinkEscape = false;
	let lastRow = rows.at(-1);
	let visible = lastRow === void 0 ? 0 : fastStringWidth(lastRow);
	let currentCharacter = characters.next();
	let nextCharacter = characters.next();
	let rawCharacterIndex = 0;
	while (!currentCharacter.done) {
		const character = currentCharacter.value;
		const characterLength = fastStringWidth(character);
		if (visible + characterLength <= columns) rows[rows.length - 1] += character;
		else {
			rows.push(character);
			visible = 0;
		}
		if (character === ESC || character === CSI) {
			isInsideEscape = true;
			isInsideLinkEscape = word.startsWith(ANSI_ESCAPE_LINK, rawCharacterIndex + 1);
		}
		if (isInsideEscape) {
			if (isInsideLinkEscape) {
				if (character === ANSI_ESCAPE_BELL) {
					isInsideEscape = false;
					isInsideLinkEscape = false;
				}
			} else if (character === ANSI_SGR_TERMINATOR) isInsideEscape = false;
		} else {
			visible += characterLength;
			if (visible === columns && !nextCharacter.done) {
				rows.push("");
				visible = 0;
			}
		}
		currentCharacter = nextCharacter;
		nextCharacter = characters.next();
		rawCharacterIndex += character.length;
	}
	lastRow = rows.at(-1);
	if (!visible && lastRow !== void 0 && lastRow.length && rows.length > 1) rows[rows.length - 2] += rows.pop();
};
const stringVisibleTrimSpacesRight = (string) => {
	const words = string.split(" ");
	let last = words.length;
	while (last) {
		if (fastStringWidth(words[last - 1])) break;
		last--;
	}
	if (last === words.length) return string;
	return words.slice(0, last).join(" ") + words.slice(last).join("");
};
const exec = (string, columns, options = {}) => {
	if (options.trim !== false && string.trim() === "") return "";
	let returnValue = "";
	let escapeCode;
	let escapeUrl;
	const words = string.split(" ");
	let rows = [""];
	let rowLength = 0;
	for (let index = 0; index < words.length; index++) {
		const word = words[index];
		if (options.trim !== false) {
			const row = rows.at(-1) ?? "";
			const trimmed = row.trimStart();
			if (row.length !== trimmed.length) {
				rows[rows.length - 1] = trimmed;
				rowLength = fastStringWidth(trimmed);
			}
		}
		if (index !== 0) {
			if (rowLength >= columns && (options.wordWrap === false || options.trim === false)) {
				rows.push("");
				rowLength = 0;
			}
			if (rowLength || options.trim === false) {
				rows[rows.length - 1] += " ";
				rowLength++;
			}
		}
		const wordLength = fastStringWidth(word);
		if (options.hard && wordLength > columns) {
			const remainingColumns = columns - rowLength;
			const breaksStartingThisLine = 1 + Math.floor((wordLength - remainingColumns - 1) / columns);
			if (Math.floor((wordLength - 1) / columns) < breaksStartingThisLine) rows.push("");
			wrapWord(rows, word, columns);
			rowLength = fastStringWidth(rows.at(-1) ?? "");
			continue;
		}
		if (rowLength + wordLength > columns && rowLength && wordLength) {
			if (options.wordWrap === false && rowLength < columns) {
				wrapWord(rows, word, columns);
				rowLength = fastStringWidth(rows.at(-1) ?? "");
				continue;
			}
			rows.push("");
			rowLength = 0;
		}
		if (rowLength + wordLength > columns && options.wordWrap === false) {
			wrapWord(rows, word, columns);
			rowLength = fastStringWidth(rows.at(-1) ?? "");
			continue;
		}
		rows[rows.length - 1] += word;
		rowLength += wordLength;
	}
	if (options.trim !== false) rows = rows.map((row) => stringVisibleTrimSpacesRight(row));
	const preString = rows.join("\n");
	let inSurrogate = false;
	for (let i = 0; i < preString.length; i++) {
		const character = preString[i];
		returnValue += character;
		if (!inSurrogate) {
			inSurrogate = character >= "\ud800" && character <= "\udbff";
			if (inSurrogate) continue;
		} else inSurrogate = false;
		if (character === ESC || character === CSI) {
			GROUP_REGEX.lastIndex = i + 1;
			const groups = GROUP_REGEX.exec(preString)?.groups;
			if (groups?.code !== void 0) {
				const code = Number.parseFloat(groups.code);
				escapeCode = code === END_CODE ? void 0 : code;
			} else if (groups?.uri !== void 0) escapeUrl = groups.uri.length === 0 ? void 0 : groups.uri;
		}
		if (preString[i + 1] === "\n") {
			if (escapeUrl) returnValue += wrapAnsiHyperlink("");
			const closingCode = escapeCode ? getClosingCode(escapeCode) : void 0;
			if (escapeCode && closingCode) returnValue += wrapAnsiCode(closingCode);
		} else if (character === "\n") {
			if (escapeCode && getClosingCode(escapeCode)) returnValue += wrapAnsiCode(escapeCode);
			if (escapeUrl) returnValue += wrapAnsiHyperlink(escapeUrl);
		}
	}
	return returnValue;
};
const CRLF_OR_LF = /\r?\n/;
function wrapAnsi(string, columns, options) {
	return String(string).normalize().split(CRLF_OR_LF).map((line) => exec(line, columns, options)).join("\n");
}
//#endregion
//#region ../../node_modules/.pnpm/@clack+core@1.4.2/node_modules/@clack/core/dist/index.mjs
var import_src = (/* @__PURE__ */ __commonJSMin(((exports, module) => {
	const ESC = "\x1B";
	const CSI = `${ESC}[`;
	const beep = "\x07";
	const cursor = {
		to(x, y) {
			if (!y) return `${CSI}${x + 1}G`;
			return `${CSI}${y + 1};${x + 1}H`;
		},
		move(x, y) {
			let ret = "";
			if (x < 0) ret += `${CSI}${-x}D`;
			else if (x > 0) ret += `${CSI}${x}C`;
			if (y < 0) ret += `${CSI}${-y}A`;
			else if (y > 0) ret += `${CSI}${y}B`;
			return ret;
		},
		up: (count = 1) => `${CSI}${count}A`,
		down: (count = 1) => `${CSI}${count}B`,
		forward: (count = 1) => `${CSI}${count}C`,
		backward: (count = 1) => `${CSI}${count}D`,
		nextLine: (count = 1) => `${CSI}E`.repeat(count),
		prevLine: (count = 1) => `${CSI}F`.repeat(count),
		left: `${CSI}G`,
		hide: `${CSI}?25l`,
		show: `${CSI}?25h`,
		save: `${ESC}7`,
		restore: `${ESC}8`
	};
	module.exports = {
		cursor,
		scroll: {
			up: (count = 1) => `${CSI}S`.repeat(count),
			down: (count = 1) => `${CSI}T`.repeat(count)
		},
		erase: {
			screen: `${CSI}2J`,
			up: (count = 1) => `${CSI}1J`.repeat(count),
			down: (count = 1) => `${CSI}J`.repeat(count),
			line: `${CSI}2K`,
			lineEnd: `${CSI}K`,
			lineStart: `${CSI}1K`,
			lines(count) {
				let clear = "";
				for (let i = 0; i < count; i++) clear += this.line + (i < count - 1 ? cursor.up() : "");
				if (count) clear += cursor.left;
				return clear;
			}
		},
		beep
	};
})))();
const settings = {
	actions: /* @__PURE__ */ new Set([
		"up",
		"down",
		"left",
		"right",
		"space",
		"enter",
		"cancel"
	]),
	aliases: /* @__PURE__ */ new Map([
		["k", "up"],
		["j", "down"],
		["h", "left"],
		["l", "right"],
		["", "cancel"],
		["escape", "cancel"]
	]),
	messages: {
		cancel: "Canceled",
		error: "Something went wrong"
	},
	withGuide: true,
	date: {
		monthNames: [...[
			"January",
			"February",
			"March",
			"April",
			"May",
			"June",
			"July",
			"August",
			"September",
			"October",
			"November",
			"December"
		]],
		messages: {
			required: "Please enter a valid date",
			invalidMonth: "There are only 12 months in a year",
			invalidDay: (n, e) => `There are only ${n} days in ${e}`,
			afterMin: (n) => `Date must be on or after ${n.toISOString().slice(0, 10)}`,
			beforeMax: (n) => `Date must be on or before ${n.toISOString().slice(0, 10)}`
		}
	}
};
function isActionKey(n, e) {
	if (typeof n == "string") return settings.aliases.get(n) === e;
	for (const s of n) if (s !== void 0 && isActionKey(s, e)) return true;
	return false;
}
function diffLines(i, s) {
	if (i === s) return;
	const e = i.split(`
`), t = s.split(`
`), r = Math.max(e.length, t.length), f = [];
	for (let n = 0; n < r; n++) e[n] !== t[n] && f.push(n);
	return {
		lines: f,
		numLinesBefore: e.length,
		numLinesAfter: t.length,
		numLines: r
	};
}
globalThis.process.platform.startsWith("win");
const CANCEL_SYMBOL = Symbol("clack:cancel");
function isCancel(e) {
	return e === CANCEL_SYMBOL;
}
function setRawMode(e, r) {
	const o = e;
	o.isTTY && o.setRawMode(r);
}
const getColumns = (e) => "columns" in e && typeof e.columns == "number" ? e.columns : 80, getRows = (e) => "rows" in e && typeof e.rows == "number" ? e.rows : 20;
function wrapTextWithPrefix(e, r, o, t = o, s = o, n) {
	return wrapAnsi(r, getColumns(e ?? stdout) - o.length, {
		hard: true,
		trim: false
	}).split(`
`).map((c, i, m) => {
		const d = n ? n(c, i) : c;
		return i === 0 ? `${t}${d}` : i === m.length - 1 ? `${s}${d}` : `${o}${d}`;
	}).join(`
`);
}
function runValidation(e, n) {
	if ("~standard" in e) {
		const a = e["~standard"].validate(n);
		if (a instanceof Promise) throw new TypeError("Schema validation must be synchronous. Update `validate()` and remove any asynchronous logic.");
		return a.issues?.at(0)?.message;
	}
	return e(n);
}
var V = class {
	input;
	output;
	_abortSignal;
	rl;
	opts;
	_render;
	_track = false;
	_prevFrame = "";
	_subscribers = /* @__PURE__ */ new Map();
	_cursor = 0;
	state = "initial";
	error = "";
	value;
	userInput = "";
	constructor(t, e = true) {
		const { input: i = stdin, output: n = stdout, render: s, signal: r, ...o } = t;
		this.opts = o, this.onKeypress = this.onKeypress.bind(this), this.close = this.close.bind(this), this.render = this.render.bind(this), this._render = s.bind(this), this._track = e, this._abortSignal = r, this.input = i, this.output = n;
	}
	/**
	* Unsubscribe all listeners
	*/
	unsubscribe() {
		this._subscribers.clear();
	}
	/**
	* Set a subscriber with opts
	* @param event - The event name
	*/
	setSubscriber(t, e) {
		const i = this._subscribers.get(t) ?? [];
		i.push(e), this._subscribers.set(t, i);
	}
	/**
	* Subscribe to an event
	* @param event - The event name
	* @param cb - The callback
	*/
	on(t, e) {
		this.setSubscriber(t, { cb: e });
	}
	/**
	* Subscribe to an event once
	* @param event - The event name
	* @param cb - The callback
	*/
	once(t, e) {
		this.setSubscriber(t, {
			cb: e,
			once: true
		});
	}
	/**
	* Emit an event with data
	* @param event - The event name
	* @param data - The data to pass to the callback
	*/
	emit(t, ...e) {
		const i = this._subscribers.get(t) ?? [], n = [];
		for (const s of i) s.cb(...e), s.once && n.push(() => i.splice(i.indexOf(s), 1));
		for (const s of n) s();
	}
	prompt() {
		return new Promise((t) => {
			if (this._abortSignal) {
				if (this._abortSignal.aborted) return this.state = "cancel", this.close(), t(CANCEL_SYMBOL);
				this._abortSignal.addEventListener("abort", () => {
					this.state = "cancel", this.close();
				}, { once: true });
			}
			this.rl = l__default.createInterface({
				input: this.input,
				tabSize: 2,
				prompt: "",
				escapeCodeTimeout: 50,
				terminal: true
			}), this.rl.prompt(), this.opts.initialUserInput !== void 0 && this._setUserInput(this.opts.initialUserInput, true), this.input.on("keypress", this.onKeypress), setRawMode(this.input, true), this.output.on("resize", this.render), this.render(), this.once("submit", () => {
				this.output.write(import_src.cursor.show), this.output.off("resize", this.render), setRawMode(this.input, false), t(this.value);
			}), this.once("cancel", () => {
				this.output.write(import_src.cursor.show), this.output.off("resize", this.render), setRawMode(this.input, false), t(CANCEL_SYMBOL);
			});
		});
	}
	_isActionKey(t, e) {
		return t === "	";
	}
	_shouldSubmit(t, e) {
		return true;
	}
	_setValue(t) {
		this.value = t, this.emit("value", this.value);
	}
	_setUserInput(t, e) {
		this.userInput = t ?? "", this.emit("userInput", this.userInput), e && this._track && this.rl && (this.rl.write(this.userInput), this._cursor = this.rl.cursor);
	}
	_clearUserInput() {
		this.rl?.write(null, {
			ctrl: true,
			name: "u"
		}), this._setUserInput("");
	}
	onKeypress(t, e) {
		if (this._track && e.name !== "return" && (e.name && this._isActionKey(t, e) && this.rl?.write(null, {
			ctrl: true,
			name: "h"
		}), this._cursor = this.rl?.cursor ?? 0, this._setUserInput(this.rl?.line)), this.state === "error" && (this.state = "active"), e?.name && (!this._track && settings.aliases.has(e.name) && this.emit("cursor", settings.aliases.get(e.name)), settings.actions.has(e.name) && this.emit("cursor", e.name)), t && (t.toLowerCase() === "y" || t.toLowerCase() === "n") && this.emit("confirm", t.toLowerCase() === "y"), this.emit("key", t, e), e?.name === "return" && this._shouldSubmit(t, e)) {
			if (this.opts.validate) {
				const i = runValidation(this.opts.validate, this.value);
				i && (this.error = i instanceof Error ? i.message : i, this.state = "error", this.rl?.write(this.userInput));
			}
			this.state !== "error" && (this.state = "submit");
		}
		isActionKey([
			t,
			e?.name,
			e?.sequence
		], "cancel") && (this.state = "cancel"), (this.state === "submit" || this.state === "cancel") && this.emit("finalize"), this.render(), (this.state === "submit" || this.state === "cancel") && this.close();
	}
	close() {
		this.input.unpipe(), this.input.removeListener("keypress", this.onKeypress), this.output.write(`
`), setRawMode(this.input, false), this.rl?.close(), this.rl = void 0, this.emit(`${this.state}`, this.value), this.unsubscribe();
	}
	restoreCursor() {
		const t = wrapAnsi(this._prevFrame, process.stdout.columns, {
			hard: true,
			trim: false
		}).split(`
`).length - 1;
		this.output.write(import_src.cursor.move(-999, t * -1));
	}
	render() {
		const t = wrapAnsi(this._render(this) ?? "", process.stdout.columns, {
			hard: true,
			trim: false
		});
		if (t !== this._prevFrame) {
			if (this.state === "initial") this.output.write(import_src.cursor.hide);
			else {
				const e = diffLines(this._prevFrame, t), i = getRows(this.output);
				if (this.restoreCursor(), e) {
					const n = Math.max(0, e.numLinesAfter - i), s = Math.max(0, e.numLinesBefore - i);
					let r = e.lines.find((o) => o >= n);
					if (r === void 0) {
						this._prevFrame = t;
						return;
					}
					if (e.lines.length === 1) {
						this.output.write(import_src.cursor.move(0, r - s)), this.output.write(import_src.erase.lines(1));
						const o = t.split(`
`);
						this.output.write(o[r]), this._prevFrame = t, this.output.write(import_src.cursor.move(0, o.length - r - 1));
						return;
					} else if (e.lines.length > 1) {
						if (n < s) r = n;
						else {
							const h = r - s;
							h > 0 && this.output.write(import_src.cursor.move(0, h));
						}
						this.output.write(import_src.erase.down());
						const f = t.split(`
`).slice(r);
						this.output.write(f.join(`
`)), this._prevFrame = t;
						return;
					}
				}
				this.output.write(import_src.erase.down());
			}
			this.output.write(t), this.state === "initial" && (this.state = "active"), this._prevFrame = t;
		}
	}
};
var r = class extends V {
	get cursor() {
		return this.value ? 0 : 1;
	}
	get _value() {
		return this.cursor === 0;
	}
	constructor(t) {
		super(t, false), this.value = !!t.initialValue, this.on("userInput", () => {
			this.value = this._value;
		}), this.on("confirm", (i) => {
			this.output.write(import_src.cursor.move(0, -1)), this.value = i, this.state = "submit", this.close();
		}), this.on("cursor", () => {
			this.value = !this.value;
		});
	}
};
//#endregion
//#region ../../node_modules/.pnpm/@clack+prompts@1.6.0/node_modules/@clack/prompts/dist/index.mjs
function isUnicodeSupported() {
	if (process$1.platform !== "win32") return process$1.env.TERM !== "linux";
	return Boolean(process$1.env.CI) || Boolean(process$1.env.WT_SESSION) || Boolean(process$1.env.TERMINUS_SUBLIME) || process$1.env.ConEmuTask === "{cmd::Cmder}" || process$1.env.TERM_PROGRAM === "Terminus-Sublime" || process$1.env.TERM_PROGRAM === "vscode" || process$1.env.TERM === "xterm-256color" || process$1.env.TERM === "alacritty" || process$1.env.TERMINAL_EMULATOR === "JetBrains-JediTerm";
}
const unicode = isUnicodeSupported(), unicodeOr = (o, e) => unicode ? o : e, S_STEP_ACTIVE = unicodeOr("◆", "*"), S_STEP_CANCEL = unicodeOr("■", "x"), S_STEP_ERROR = unicodeOr("▲", "x"), S_STEP_SUBMIT = unicodeOr("◇", "o"), S_BAR = unicodeOr("│", "|"), S_BAR_END = unicodeOr("└", "—"), S_RADIO_ACTIVE = unicodeOr("●", ">"), S_RADIO_INACTIVE = unicodeOr("○", " "), S_BAR_H = unicodeOr("─", "-"), S_CORNER_TOP_RIGHT = unicodeOr("╮", "+"), S_CONNECT_LEFT = unicodeOr("├", "+"), S_CORNER_BOTTOM_RIGHT = unicodeOr("╯", "+"), S_CORNER_BOTTOM_LEFT = unicodeOr("╰", "+"), symbol = (o) => {
	switch (o) {
		case "initial":
		case "active": return styleText("cyan", S_STEP_ACTIVE);
		case "cancel": return styleText("red", S_STEP_CANCEL);
		case "error": return styleText("yellow", S_STEP_ERROR);
		case "submit": return styleText("green", S_STEP_SUBMIT);
	}
};
const confirm = (i) => {
	const a = i.active ?? "Yes", s = i.inactive ?? "No";
	return new r({
		active: a,
		inactive: s,
		signal: i.signal,
		input: i.input,
		output: i.output,
		initialValue: i.initialValue ?? true,
		render() {
			const e = i.withGuide ?? settings.withGuide, u = `${symbol(this.state)}  `, l = e ? `${styleText("gray", S_BAR)}  ` : "", f = wrapTextWithPrefix(i.output, i.message, l, u), o = `${e ? `${styleText("gray", S_BAR)}
` : ""}${f}
`, c = this.value ? a : s;
			switch (this.state) {
				case "submit": return `${o}${e ? `${styleText("gray", S_BAR)}  ` : ""}${styleText("dim", c)}`;
				case "cancel": return `${o}${e ? `${styleText("gray", S_BAR)}  ` : ""}${styleText(["strikethrough", "dim"], c)}${e ? `
${styleText("gray", S_BAR)}` : ""}`;
				default: {
					const r = e ? `${styleText("cyan", S_BAR)}  ` : "", g = e ? styleText("cyan", S_BAR_END) : "";
					return `${o}${r}${this.value ? `${styleText("green", S_RADIO_ACTIVE)} ${a}` : `${styleText("dim", S_RADIO_INACTIVE)} ${styleText("dim", a)}`}${i.vertical ? e ? `
${styleText("cyan", S_BAR)}  ` : `
` : ` ${styleText("dim", "/")} `}${this.value ? `${styleText("dim", S_RADIO_INACTIVE)} ${styleText("dim", s)}` : `${styleText("green", S_RADIO_ACTIVE)} ${s}`}
${g}
`;
				}
			}
		}
	}).prompt();
};
`${styleText("dim", "↑/↓")}`, `${styleText("dim", "Space:")}`, `${styleText("dim", "Enter:")}`;
const outro = (o = "", t) => {
	const i = t?.output ?? process.stdout, e = t?.withGuide ?? settings.withGuide ? `${styleText("gray", S_BAR)}
${styleText("gray", S_BAR_END)}  ` : "";
	i.write(`${e}${o}

`);
};
const W$1 = (o) => o, C = (o, e, s) => {
	const a = {
		hard: true,
		trim: false
	}, i = wrapAnsi(o, e, a).split(`
`), c = i.reduce((n, t) => Math.max(fastStringWidth(t), n), 0);
	return wrapAnsi(o, e - (i.map(s).reduce((n, t) => Math.max(fastStringWidth(t), n), 0) - c), a);
};
const note = (o = "", e = "", s) => {
	const a = s?.output ?? process$1.stdout, i = s?.withGuide ?? settings.withGuide, c = s?.format ?? W$1, g = [
		"",
		...C(o, getColumns(a) - 6, c).split(`
`).map(c),
		""
	], n = fastStringWidth(e), t = Math.max(g.reduce((m, F) => {
		const O = fastStringWidth(F);
		return O > m ? O : m;
	}, 0), n) + 2, h = g.map((m) => `${styleText("gray", S_BAR)}  ${m}${" ".repeat(t - fastStringWidth(m))}${styleText("gray", S_BAR)}`).join(`
`), T = i ? `${styleText("gray", S_BAR)}
` : "", l$1 = i ? S_CONNECT_LEFT : S_CORNER_BOTTOM_LEFT;
	a.write(`${T}${styleText("green", S_STEP_SUBMIT)}  ${styleText("reset", e)} ${styleText("gray", S_BAR_H.repeat(Math.max(t - n - 1, 1)) + S_CORNER_TOP_RIGHT)}
${h}
${styleText("gray", l$1 + S_BAR_H.repeat(t + 2) + S_CORNER_BOTTOM_RIGHT)}
`);
};
`${styleText("dim", "↑/↓")}`, `${styleText("dim", "Enter:")}`;
`${styleText("gray", S_BAR)}`;
//#endregion
//#region src/node/rpc/anonymous/auth.ts
const AUTH_TIMEOUT_MS = 6e4;
const anonymousAuth = defineRpcFunction({
	name: "devframe:anonymous:auth",
	type: "action",
	jsonSerializable: true,
	setup: (context) => {
		const storage = getInternalContext(context).storage.auth;
		return { handler: async (query) => {
			const session = context.rpc.getCurrentRpcSession();
			if (!session) throw new Error("Failed to retrieve the current RPC session");
			if (session.meta.isTrusted || storage.value().trusted[query.authToken]) {
				session.meta.clientAuthToken = query.authToken;
				session.meta.isTrusted = true;
				return { isTrusted: true };
			}
			if ((context.viteConfig.devtools?.config?.clientAuthTokens ?? []).includes(query.authToken)) {
				session.meta.clientAuthToken = query.authToken;
				session.meta.isTrusted = true;
				return { isTrusted: true };
			}
			if (query.authToken === getTempAuthToken()) {
				storage.mutate((state) => {
					state.trusted[query.authToken] = {
						authToken: query.authToken,
						ua: query.ua,
						origin: query.origin,
						timestamp: Date.now()
					};
				});
				session.meta.clientAuthToken = query.authToken;
				session.meta.isTrusted = true;
				refreshTempAuthToken();
				return { isTrusted: true };
			}
			abortPendingAuth();
			const tempId = getTempAuthToken();
			const authUrl = `${context.viteServer?.resolvedUrls?.local?.[0]?.replace(/\/$/, "") ?? `http://localhost:${context.viteConfig.server.port}`}/__devtools/auth?id=${encodeURIComponent(tempId)}`;
			const message = [
				`A browser is requesting permissions to connect to the Vite DevTools.`,
				"",
				`User Agent   : ${colors.yellow(colors.bold(query.ua || "Unknown"))}`,
				`Origin       : ${colors.yellow(colors.bold(query.origin || "Unknown"))}`,
				`Client Token : ${colors.green(colors.bold(query.authToken))}`,
				"",
				`Manual Auth URL   : ${colors.cyan(colors.underline(authUrl))}`,
				`Manual Auth Token : ${colors.cyan(colors.bold(tempId))}`,
				"",
				"This will allow the browser to interact with the server, make file changes and run commands.",
				colors.red(colors.bold("You should only trust your local development browsers."))
			];
			note(colors.reset(message.join("\n")), colors.bold(colors.yellow(" Vite DevTools Permission Request ")));
			if (!process$1.stdout.isTTY) return { isTrusted: false };
			const abortController = new AbortController();
			return new Promise((resolve) => {
				const timeout = setTimeout(() => {
					abortController.abort();
					setPendingAuth(null);
					console.log(colors.yellow`${MARK_INFO} Auth request timed out for ${colors.bold(query.authToken)}`);
					resolve({ isTrusted: false });
				}, AUTH_TIMEOUT_MS);
				setPendingAuth({
					clientAuthToken: query.authToken,
					session,
					ua: query.ua,
					origin: query.origin,
					resolve,
					abortController,
					timeout
				});
				confirm({
					message: colors.bold(`Do you trust this client (${colors.green(colors.bold(query.authToken))})?`),
					initialValue: false,
					signal: abortController.signal
				}).then((answer) => {
					clearTimeout(timeout);
					setPendingAuth(null);
					if (isCancel(answer)) return;
					if (answer) {
						storage.mutate((state) => {
							state.trusted[query.authToken] = {
								authToken: query.authToken,
								ua: query.ua,
								origin: query.origin,
								timestamp: Date.now()
							};
						});
						session.meta.clientAuthToken = query.authToken;
						session.meta.isTrusted = true;
						outro(colors.green(colors.bold(`You have granted permissions to ${colors.bold(query.authToken)}`)));
						resolve({ isTrusted: true });
					} else {
						outro(colors.red(colors.bold(`You have denied permissions to ${colors.bold(query.authToken)}`)));
						resolve({ isTrusted: false });
					}
				}).catch(() => {
					clearTimeout(timeout);
					setPendingAuth(null);
				});
			});
		} };
	}
});
//#endregion
//#region src/node/rpc/internal/commands-execute.ts
const commandsExecute = defineRpcFunction({
	name: "devtoolskit:internal:commands:execute",
	type: "action",
	setup: (context) => {
		return { async handler(id, ...args) {
			return context.commands.execute(id, ...args);
		} };
	}
});
//#endregion
//#region src/node/rpc/internal/commands-list.ts
const commandsList = defineRpcFunction({
	name: "devtoolskit:internal:commands:list",
	type: "static",
	jsonSerializable: true,
	setup: (context) => {
		return { async handler() {
			return context.commands.list();
		} };
	}
});
//#endregion
//#region src/node/rpc/internal/docks-on-launch.ts
const docksOnLaunch = defineRpcFunction({
	name: "devtoolskit:internal:docks:on-launch",
	type: "action",
	setup: (context) => {
		const launchMap = /* @__PURE__ */ new Map();
		return { handler: async (entryId) => {
			if (launchMap.has(entryId)) return launchMap.get(entryId);
			const entry = context.docks.values().find((entry) => entry.id === entryId);
			if (!entry) throw diagnostics.DTK0030({ id: entryId });
			if (entry.type !== "launcher") throw diagnostics.DTK0031({ id: entryId });
			try {
				context.docks.update({
					...entry,
					launcher: {
						...entry.launcher,
						status: "loading"
					}
				});
				const promise = entry.launcher.onLaunch();
				launchMap.set(entryId, promise);
				const result = await promise;
				const newEntry = context.docks.values().find((entry) => entry.id === entryId) || entry;
				if (newEntry.type === "launcher") context.docks.update({
					...newEntry,
					launcher: {
						...newEntry.launcher,
						status: "success"
					}
				});
				return result;
			} catch (error) {
				diagnostics.DTK0032({
					id: entryId,
					cause: error
				});
				context.docks.update({
					...entry,
					launcher: {
						...entry.launcher,
						status: "error",
						error: error instanceof Error ? error.message : String(error)
					}
				});
			}
		} };
	}
});
//#endregion
//#region src/node/rpc/internal/messages-add.ts
const messagesAdd = defineRpcFunction({
	name: "devtoolskit:internal:messages:add",
	type: "action",
	jsonSerializable: true,
	setup: (context) => {
		return { async handler(input) {
			return (await context.messages.add({
				...input,
				from: "browser"
			})).entry;
		} };
	}
});
//#endregion
//#region src/node/rpc/internal/messages-clear.ts
const messagesClear = defineRpcFunction({
	name: "devtoolskit:internal:messages:clear",
	type: "action",
	jsonSerializable: true,
	setup: (context) => {
		return { async handler() {
			await context.messages.clear();
		} };
	}
});
//#endregion
//#region src/node/rpc/internal/messages-list.ts
const messagesList = defineRpcFunction({
	name: "devtoolskit:internal:messages:list",
	type: "static",
	jsonSerializable: true,
	setup: (context) => {
		const host = context.messages;
		return { async handler(since) {
			const currentVersion = host._clock;
			if (since == null) return {
				entries: Array.from(host.entries.values()),
				removedIds: [],
				version: currentVersion
			};
			const entries = [];
			for (const [id, entry] of host.entries) {
				const mod = host.lastModified?.get(id);
				if (mod != null && mod > since) entries.push(entry);
			}
			const removedIds = [];
			const removals = host.removals;
			for (const r of removals) if (r.time > since) removedIds.push(r.id);
			const pruneThreshold = since;
			while (removals.length > 0 && removals[0].time <= pruneThreshold) removals.shift();
			return {
				entries,
				removedIds,
				version: currentVersion
			};
		} };
	}
});
//#endregion
//#region src/node/rpc/internal/messages-remove.ts
const messagesRemove = defineRpcFunction({
	name: "devtoolskit:internal:messages:remove",
	type: "action",
	jsonSerializable: true,
	setup: (context) => {
		return { async handler(id) {
			await context.messages.remove(id);
		} };
	}
});
//#endregion
//#region src/node/rpc/internal/messages-update.ts
const messagesUpdate = defineRpcFunction({
	name: "devtoolskit:internal:messages:update",
	type: "action",
	jsonSerializable: true,
	setup: (context) => {
		return { async handler(id, patch) {
			return await context.messages.update(id, patch) ?? null;
		} };
	}
});
//#endregion
//#region src/node/rpc/internal/rpc-server-list.ts
const rpcServerList = defineRpcFunction({
	name: "devtoolskit:internal:rpc:server:list",
	type: "static",
	jsonSerializable: true,
	setup: (context) => {
		return { async handler() {
			return Object.fromEntries(Array.from(context.rpc.definitions.entries()).map(([name, fn]) => [name, {
				type: fn.type,
				cacheable: fn.cacheable || false
			}]));
		} };
	}
});
//#endregion
//#region src/node/rpc/internal/terminals-list.ts
const terminalsList = defineRpcFunction({
	name: "devtoolskit:internal:terminals:list",
	type: "static",
	jsonSerializable: true,
	setup: (context) => {
		return { async handler() {
			return Array.from(context.terminals.sessions.values()).map((i) => {
				return {
					id: i.id,
					title: i.title,
					description: i.description,
					status: i.status
				};
			});
		} };
	}
});
//#endregion
//#region src/node/rpc/internal/terminals-read.ts
const terminalsRead = defineRpcFunction({
	name: "devtoolskit:internal:terminals:read",
	type: "query",
	jsonSerializable: true,
	setup: (context) => {
		return { async handler(id) {
			const session = context.terminals.sessions.get(id);
			if (!session) throw new Error(`Terminal session with id "${id}" not found`);
			return {
				buffer: session.buffer ?? [],
				ts: Date.now()
			};
		} };
	}
});
//#endregion
//#region src/node/rpc/index.ts
const builtinPublicRpcDeclarations = [defineRpcFunction({
	name: "vite:core:open-in-editor",
	type: "action",
	jsonSerializable: true,
	setup: (context) => {
		return { handler: async (path) => {
			const resolved = resolve(context.workspaceRoot, path);
			const rel = relative(context.workspaceRoot, resolved);
			if (rel.startsWith("..") || rel.includes("\0")) throw diagnostics.DTK0028();
			launchEditor(resolved);
		} };
	}
}), defineRpcFunction({
	name: "vite:core:open-in-finder",
	type: "action",
	jsonSerializable: true,
	setup: (context) => {
		return { handler: async (path) => {
			const resolved = resolve(context.workspaceRoot, path);
			const rel = relative(context.workspaceRoot, resolved);
			if (rel.startsWith("..") || rel.includes("\0")) throw diagnostics.DTK0029();
			await open(resolved);
		} };
	}
})];
const builtinAnonymousRpcDeclarations = [anonymousAuth];
const builtinInternalRpcDeclarations = [
	commandsExecute,
	commandsList,
	docksOnLaunch,
	messagesAdd,
	messagesClear,
	messagesList,
	messagesRemove,
	messagesUpdate,
	rpcServerList,
	terminalsList,
	terminalsRead
];
const builtinRpcDeclarations = [
	...builtinPublicRpcDeclarations,
	...builtinAnonymousRpcDeclarations,
	...builtinInternalRpcDeclarations
];
//#endregion
//#region src/node/context.ts
var context_exports = /* @__PURE__ */ __exportAll({ createDevToolsContext: () => createDevToolsContext });
const debugSetup = createDebug("vite:devtools:context:setup");
function shouldSkipSetupByCapabilities(plugin, mode) {
	const modeCapabilities = plugin.devtools?.capabilities?.[mode];
	if (modeCapabilities === false) return true;
	if (!isObject(modeCapabilities)) return false;
	return Object.values(modeCapabilities).includes(false);
}
async function createDevToolsContext(viteConfig, viteServer) {
	const cwd = viteConfig.root;
	const { searchForWorkspaceRoot } = await import("vite");
	const mode = viteConfig.command === "serve" ? "dev" : "build";
	const workspaceRoot = searchForWorkspaceRoot(cwd) ?? cwd;
	const context = await createKitContext({
		cwd,
		workspaceRoot,
		mode,
		host: createViteDevToolsHost({
			viteConfig,
			viteServer,
			workspaceRoot
		}),
		builtinRpcDeclarations,
		viteConfig,
		viteServer
	});
	context.diagnostics.register(diagnostics);
	const rpcHost = context.rpc;
	context.commands.register({
		id: "vite:open-in-editor",
		title: "Open in Editor",
		icon: "ph:pencil-duotone",
		category: "editor",
		showInPalette: false,
		handler: (path) => rpcHost.invokeLocal("vite:core:open-in-editor", path)
	});
	context.commands.register({
		id: "vite:open-in-finder",
		title: "Open in Finder",
		icon: "ph:folder-open-duotone",
		category: "editor",
		showInPalette: false,
		handler: (path) => rpcHost.invokeLocal("vite:core:open-in-finder", path)
	});
	context.docks.register({
		id: DEVTOOLS_VITEPLUS_GROUP_ID,
		type: "group",
		title: "Vite+",
		icon: {
			light: "builtin:vite-plus-core",
			dark: "builtin:vite-plus-core"
		},
		defaultOrder: -1e3
	});
	const plugins = viteConfig.plugins.filter((plugin) => "devtools" in plugin);
	for (const plugin of plugins) {
		if (!plugin.devtools?.setup) continue;
		if (shouldSkipSetupByCapabilities(plugin, mode)) {
			debugSetup(`skipping plugin ${JSON.stringify(plugin.name)} due to disabled capabilities in ${mode} mode`);
			continue;
		}
		try {
			debugSetup(`setting up plugin ${JSON.stringify(plugin.name)}`);
			await plugin.devtools?.setup?.(context);
		} catch (error) {
			throw diagnostics.DTK0014({
				name: plugin.name,
				cause: error
			});
		}
	}
	return context;
}
//#endregion
//#region src/node/plugins/server.ts
var server_exports = /* @__PURE__ */ __exportAll({
	DevToolsServer: () => DevToolsServer,
	renderDockImportsMap: () => renderDockImportsMap
});
/**
* Core plugin for enabling Vite DevTools
*/
function renderDockImportsMap(docks) {
	const map = /* @__PURE__ */ new Map();
	for (const dock of docks) {
		const id = `${dock.type}:${dock.id}`;
		if (dock.type === "action") map.set(id, dock.action);
		else if (dock.type === "custom-render") map.set(id, dock.renderer);
		else if (dock.type === "iframe" && dock.clientScript) map.set(id, dock.clientScript);
	}
	return [
		`export const importsMap = {`,
		...[...map.entries()].filter(([, entry]) => entry != null).map(([id, { importFrom, importName }]) => `  [${JSON.stringify(id)}]: () => import(${JSON.stringify(importFrom)}).then(r => r[${JSON.stringify(importName ?? "default")}]),`),
		"}"
	].join("\n");
}
function DevToolsServer() {
	let context;
	return {
		name: "vite:devtools:server",
		enforce: "post",
		apply: "serve",
		async configureServer(viteDevServer) {
			context = await createDevToolsContext(viteDevServer.config, viteDevServer);
			const host = viteDevServer.config.server.host === true ? "0.0.0.0" : viteDevServer.config.server.host || "localhost";
			const { middleware } = await createDevToolsMiddleware({
				cwd: viteDevServer.config.root,
				websocket: { host },
				context
			});
			viteDevServer.middlewares.use((req, res, next) => {
				if (req.url === DEVTOOLS_MOUNT_PATH_NO_TRAILING_SLASH || req.url?.startsWith(`${DEVTOOLS_MOUNT_PATH_NO_TRAILING_SLASH}?`)) {
					res.statusCode = 302;
					res.setHeader("Location", `${DEVTOOLS_MOUNT_PATH}${req.url.slice(DEVTOOLS_MOUNT_PATH_NO_TRAILING_SLASH.length)}`);
					res.end();
					return;
				}
				next();
			});
			viteDevServer.middlewares.use(DEVTOOLS_MOUNT_PATH, middleware);
		},
		resolveId(id) {
			if (id === DEVTOOLS_DOCK_IMPORTS_VIRTUAL_ID) return id;
		},
		load(id) {
			if (id === DEVTOOLS_DOCK_IMPORTS_VIRTUAL_ID) {
				if (!context) throw new Error("DevTools context is not initialized");
				return renderDockImportsMap(context.docks.values());
			}
		}
	};
}
//#endregion
export { createDevToolsContext as i, server_exports as n, context_exports as r, DevToolsServer as t };
