// @vitejs/devtools/dist/server-CRE2OchS.js
import { i as __exportAll, n as getPort } from "./dist-HTAOCnE-.js";
import { n as diagnostics, t as MARK_INFO } from "./constants-CfdflDK5.js";
import { dirClientStandalone } from "./dirs.js";
import { DEVTOOLS_CONNECTION_META_FILENAME } from "@vitejs/devtools-kit/constants";
import { createDebug } from "obug";
import process$1 from "node:process";
import { resolve } from "node:path";
import { consumeTempAuthToken } from "devframe/node/auth";
import { getInternalContext } from "devframe/node/hub-internals";
import { colors } from "devframe/utils/colors";
import { mountStaticHandler } from "devframe/utils/serve-static";
import { H3, defineHandler, getQuery, toNodeHandler } from "h3";
import { AsyncLocalStorage } from "node:async_hooks";
import { createRpcServer } from "devframe/rpc/server";
import { attachWsRpcTransport } from "devframe/rpc/transports/ws-server";
import { readFile } from "node:fs/promises";
//#region src/node/https.ts
async function resolveHttpsConfig(https) {
	if (!https) return void 0;
	const [ca, cert, key, pfx] = await Promise.all([
		readFileIfExists(https.ca),
		readFileIfExists(https.cert),
		readFileIfExists(https.key),
		readFileIfExists(https.pfx)
	]);
	return {
		...https,
		ca,
		cert,
		key,
		pfx
	};
}
async function readFileIfExists(value) {
	if (typeof value === "string") return readFile(resolve(value)).catch(() => value);
	return value;
}
//#endregion
//#region src/node/ws.ts
const debugInvoked = createDebug("vite:devtools:rpc:invoked");
const ANONYMOUS_SCOPE = "devframe:anonymous:";
function buildWsUrl({ host, port, https }) {
	return `${https ? "wss" : "ws"}://${host === "0.0.0.0" || host === "::" || host === "" ? "localhost" : host}:${port}`;
}
async function createWsServer(options) {
	const rpcHost = options.context.rpc;
	const host = options.websocket.host;
	const https = await resolveHttpsConfig(options.websocket.https === false ? void 0 : options.websocket.https ?? options.context.viteConfig.server.https);
	const port = options.websocket.port ?? await getPort({
		port: 7812,
		host,
		random: true
	});
	const wsClients = /* @__PURE__ */ new Set();
	const context = options.context;
	const contextInternal = getInternalContext(context);
	const isClientAuthDisabled = context.mode === "build" || context.viteConfig.devtools?.config?.clientAuth === false || process$1.env.VITE_DEVTOOLS_DISABLE_CLIENT_AUTH === "true";
	if (isClientAuthDisabled) diagnostics.DTK0008();
	contextInternal.wsEndpoint = { url: buildWsUrl({
		host,
		port,
		https: !!https
	}) };
	for (const view of context.docks.views.values()) if (view.type === "iframe" && view.remote) context.docks.events.emit("dock:entry:updated", view);
	const asyncStorage = new AsyncLocalStorage();
	const rpcGroup = createRpcServer(rpcHost.functions, { rpcOptions: {
		onFunctionError(error, name) {
			diagnostics.DTK0011({
				name,
				cause: error
			});
		},
		onGeneralError(error) {
			diagnostics.DTK0012({ cause: error });
		},
		resolver(name, fn) {
			const rpc = this;
			if (!name.startsWith(ANONYMOUS_SCOPE) && !rpc.$meta.isTrusted) return () => {
				throw diagnostics.DTK0013({
					name,
					clientId: rpc.$meta.id
				});
			};
			if (!fn) return void 0;
			return async function(...args) {
				debugInvoked(`${JSON.stringify(name)} from #${rpc.$meta.id}`);
				return await asyncStorage.run({
					rpc,
					meta: rpc.$meta
				}, async () => {
					return (await fn).apply(this, args);
				});
			};
		}
	} });
	attachWsRpcTransport(rpcGroup, {
		port,
		host,
		https,
		definitions: rpcHost.definitions,
		onConnected: (ws, req, meta) => {
			const authToken = new URL(req.url ?? "", "http://localhost").searchParams.get("devframe_auth_token") ?? void 0;
			const requestOrigin = req.headers.origin;
			if (isClientAuthDisabled) meta.isTrusted = true;
			else if (authToken && contextInternal.isRemoteTokenTrusted(authToken, requestOrigin)) {
				meta.isTrusted = true;
				meta.clientAuthToken = authToken;
			} else if (authToken && contextInternal.storage.auth.value().trusted[authToken]) {
				meta.isTrusted = true;
				meta.clientAuthToken = authToken;
			} else if (authToken && (context.viteConfig.devtools?.config?.clientAuthTokens ?? []).includes(authToken)) {
				meta.isTrusted = true;
				meta.clientAuthToken = authToken;
			}
			wsClients.add(ws);
			const color = meta.isTrusted ? colors.green : colors.yellow;
			console.log(color`${MARK_INFO} Websocket client connected. [${meta.id}] [${meta.clientAuthToken}] (${meta.isTrusted ? "trusted" : "untrusted"})`);
		},
		onDisconnected: (ws, meta) => {
			wsClients.delete(ws);
			rpcHost._emitSessionDisconnected(meta);
			console.log(colors.red`${MARK_INFO} Websocket client disconnected. [${meta.id}]`);
		}
	});
	rpcHost._rpcGroup = rpcGroup;
	rpcHost._asyncStorage = asyncStorage;
	const getConnectionMeta = async () => {
		const jsonSerializableMethods = [];
		for (const def of rpcHost.definitions.values()) if (def.jsonSerializable === true) jsonSerializableMethods.push(def.name);
		return {
			backend: "websocket",
			websocket: port,
			jsonSerializableMethods
		};
	};
	return {
		port,
		rpc: rpcGroup,
		rpcHost,
		getConnectionMeta
	};
}
//#endregion
//#region src/node/server.ts
var server_exports = /* @__PURE__ */ __exportAll({ createDevToolsMiddleware: () => createDevToolsMiddleware });
function generateAuthPageHtml() {
	return `<!DOCTYPE html>
<html>
<head>
  <title>Vite DevTools Authorization</title>
  <style>
    html { font-family: system-ui, sans-serif; padding: 2rem; }
    body { height: 80vh; display: flex; flex-direction: column; justify-content: center; align-items: center; gap: 1rem; }
    #message { font-size: 1.2rem; }
    @media (prefers-color-scheme: dark) { html { background: #1a1a1a; color: #e0e0e0; } }
  </style>
</head>
<body>
  <div id="message">Verifying...</div>
  <script>
    const query = new URLSearchParams(location.search)
    const id = query.get('id')
    const el = document.getElementById('message')

    if (!id) {
      el.textContent = '\\u26a0\\ufe0f No auth token found. Please check your URL.'
      el.style.color = '#df513f'
    } else {
      fetch(location.pathname.replace(/\\/$/, '') + '-verify?id=' + encodeURIComponent(id))
        .then(async (r) => {
          if (r.status !== 200) throw new Error(await r.text())
          const data = await r.json()
          const authToken = data.authToken

          localStorage.setItem('__DEVFRAME_CONNECTION_AUTH_TOKEN__', authToken)

          try {
            const bc = new BroadcastChannel('devframe-auth')
            bc.postMessage({ type: 'auth-update', authToken: authToken })
          } catch {}

          el.textContent = '\\u2705 Authorized! You can close this window now.'
          window.close()
        })
        .catch((err) => {
          el.textContent = '\\u26a0\\ufe0f Failed to authorize: ' + err.message
          el.style.color = '#df513f'
        })
    }
  <\/script>
</body>
</html>`;
}
async function createDevToolsMiddleware(options) {
	const h3 = new H3();
	const contextInternal = getInternalContext(options.context);
	const { rpc, getConnectionMeta } = await createWsServer(options);
	h3.use(`/${DEVTOOLS_CONNECTION_META_FILENAME}`, defineHandler(async (event) => {
		event.res.headers.set("Content-Type", "application/json");
		return JSON.stringify(await getConnectionMeta());
	}));
	h3.use("/auth-verify", defineHandler((event) => {
		const { id } = getQuery(event);
		if (!id) {
			event.res.status = 400;
			return "Missing id parameter";
		}
		const clientAuthToken = consumeTempAuthToken(id, contextInternal.storage.auth);
		if (!clientAuthToken) {
			event.res.status = 403;
			return "Invalid or expired auth token";
		}
		event.res.headers.set("Content-Type", "application/json");
		return JSON.stringify({ authToken: clientAuthToken });
	}));
	h3.use("/auth", defineHandler((event) => {
		event.res.headers.set("Content-Type", "text/html; charset=utf-8");
		return generateAuthPageHtml();
	}));
	mountStaticHandler(h3, "", dirClientStandalone);
	return {
		h3,
		rpc,
		middleware: toNodeHandler(h3),
		getConnectionMeta
	};
}
//#endregion
export { server_exports as n, createDevToolsMiddleware as t };
